File: BinarySearch/C/BinarySearch.c
#include <stdio.h>

int binarySearch(int arr[], int begin, int end, int value)
{
	if(end >= begin)
	{
		int middleIndex = (begin + end) / 2;
		
		if(arr[middleIndex] == value)
			return middleIndex;
		
		if(value > arr[middleIndex])
			return binarySearch(arr, middleIndex + 1, end, value);
		else
			return binarySearch(arr, begin, middleIndex - 1, value);
			
	}	
	return -1;
}

int main()
{
	int someArr[] = {0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841};
	int sizeOfArr = sizeof(someArr) / sizeof(someArr[0]);
	
	puts("0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841");

	printf("Searched: %d, index: %d\n", 2,  binarySearch(someArr, 0, sizeOfArr - 1, 2));
	printf("Searched: %d, index: %d\n", 4,  binarySearch(someArr, 0, sizeOfArr - 1, 4));
	printf("Searched: %d, index: %d\n", 10,  binarySearch(someArr, 0, sizeOfArr - 1, 10));
	printf("Searched: %d, index: %d\n", 257,  binarySearch(someArr, 0, sizeOfArr - 1, 257));
	printf("Searched: %d, index: %d\n", 256,  binarySearch(someArr, 0, sizeOfArr - 1, 256));
	printf("Searched: %d, index: %d\n", 400,  binarySearch(someArr, 0, sizeOfArr - 1, 400));
	printf("Searched: %d, index: %d\n", 441,  binarySearch(someArr, 0, sizeOfArr - 1, 441));

	return 0;
}

File: BinarySearch/C#/BinarySearch.cs
using System;
using System.Collections.Generic;

namespace BinarySearch
{
	class Program
	{
		public static int BinarySearch(List<int> someList, int begin, int end, int val)
		{
			if(end >= begin)
			{
				int middleIndex = (begin + end) / 2;
				
				if(someList[middleIndex] == val)
					return middleIndex;

				if(val >= someList[middleIndex])
					return BinarySearch(someList, middleIndex + 1, end, val);
				else
					return BinarySearch(someList, begin, middleIndex - 1, val);
			}
			return -1;
		}

		public static void printList(List<int> anotherList)
        	{
                	foreach(int member in anotherList)
                        	Console.Write($"{member} ");
                	Console.WriteLine();
	        }

		private static void Main(string[] args)
		{
			List<int> listOne = new List<int>();
			
			for(int i = 0; i < 25; i++)
				listOne.Add(i*i);

			printList(listOne);
			Console.WriteLine("Searched for: 2, index: " + BinarySearch(listOne, 0, listOne.Count, 2));
			Console.WriteLine("Searched for: 4, index: " + BinarySearch(listOne, 0, listOne.Count, 4));
			Console.WriteLine("Searched for: 10, index: " + BinarySearch(listOne, 0, listOne.Count, 10));
			Console.WriteLine("Searched for: 257, index: " + BinarySearch(listOne, 0, listOne.Count, 257));
			Console.WriteLine("Searched for: 256, index: " + BinarySearch(listOne, 0, listOne.Count, 256));
			Console.WriteLine("Searched for: 400, index: " + BinarySearch(listOne, 0, listOne.Count, 400));
			Console.WriteLine("Searched for: 441, index: " + BinarySearch(listOne, 0, listOne.Count, 441));
		}
	}
}

